lib/autoproj/workspace.rb

```diff
@@ -457,24 +461,89 @@ def load_main_initrb
             load_if_present(local_source, config_dir, "init.rb")
         end
 
+        def self.find_path(xdg_var, xdg_path, home_path)
+            home_dir = begin Dir.home
+                       rescue ArgumentError
+                           return
+                       end
+
+            xdg_path  = File.join(XDG[xdg_var].to_path, 'autoproj', xdg_path)
+            home_path = File.join(home_dir, home_path)
+
+            if File.exist?(xdg_path)
+                xdg_path
+            elsif File.exist?(home_path)
+                home_path
+            else
+                xdg_path
+            end
+        end
+
+        def self.find_user_config_path(xdg_path, home_path = xdg_path)
+            find_path('CONFIG_HOME', xdg_path, home_path)
+        end
+
+        def self.rcfile_path
+            find_user_config_path('rc', '.autoprojrc')
+        end
+
+        def self.find_user_data_path(xdg_path, home_path = xdg_path)
+            find_path('DATA_HOME', xdg_path, File.join('.autoproj', home_path))
+        end
+
+        def self.find_user_cache_path(xdg_path, home_path = xdg_path)
+            find_path('CACHE_HOME', xdg_path, File.join('.autoproj', home_path))
+        end
+
+        RegisteredWorkspace = Struct.new :root_dir, :prefix_dir, :build_dir
+
+        def self.registered_workspaces
+            path = find_user_data_path('workspaces.yml')
+            if File.file?(path)
+                yaml = (YAML.load(File.read(path)) || [])
+                fields = RegisteredWorkspace.members.map(&:to_s)
+                yaml.map do |h|
+                    values = h.values_at(*fields)
+                    RegisteredWorkspace.new(*values)
+                end.compact
+            else
+                []
+            end
+        end
+
+        def self.save_registered_workspaces(workspaces)
+            workspaces = workspaces.map do |w|
+                Hash['root_dir' => w.root_dir,
+                     'prefix_dir' => w.prefix_dir,
```

Comment:
Is it a good idea to have this workspace-specific configuration stored globally? I kinda dislike the fact that this will probably be left outdated if the workspace's config.yml changes specially if the user has to manually register and re-register (to update) this information. I would instead keep only the `root_dir` in the global registry and load `prefix_dir` and `build_dir` from workspace's config.yml (for `autoproj global status`).

I also wonder if workspaces shouldn't be registered automatically (maybe whenever env.sh is updated?)

Reply:
It actually [is registered automatically](https://github.com/rock-core/autoproj/pull/202/files#diff-eae79889f58431d9f820a91c7210cfd8R292). The register command was there more for testing purposes. It could be removed if you believe it's better.

The point of saving the various folders it to be able to track stale folders once workspaces get deleted (for people like me that separate build, install and source).

Reply:
> It could be removed if you believe it's better

No, automatically is cool.

> The point of saving the various folders it to be able to track stale folders once workspaces get deleted (for people like me that separate build, install and source).

Makes sence.. I missed that use case.